### PR TITLE
otpilot: Implement Firmware update protocol.

### PR DESCRIPTION
**Pull request by osenft**

### CODE REVIEW

File: shared-lib/spiutils/src/compat/firmware.rs [hunk 1]
```diff
@@ -27,12 +27,17 @@ use core::mem;
 
 // ----------------------------------------------------------------------------
 
+/// The location of the BuildInfo from the start of the firmware segment.
```
Comment (dann21) [inline shared-lib/spiutils/src/compat/firmware.rs]: s/location/offset/ ?

File: userspace/otpilot/src/spi_processor.rs [hunk 2]
```diff
@@ -159,30 +169,191 @@ impl<'a> SpiProcessor<'a> {
...
+            let mut tx_cursor = SpiutilsCursor::new(&mut SPI_TX_BUF[payload::HEADER_LEN..]);
+
+            let fw_header = firmware::Header {
+                content: M::TYPE
+            };
+            fw_header.to_wire(&mut tx_cursor)?;
+            response.to_wire(&mut tx_cursor)?;
             payload_len = u16::try_from(tx_cursor.consumed_len())
-                .map_err(|_| SpiProcessorError::FromWire(FromWireError::OutOfRange))?;
+                    .map_err(|_| SpiProcessorError::FromWire(FromWireError::OutOfRange))?;
```
Comment (dann21) [inline userspace/otpilot/src/spi_processor.rs]: Nit: weird indentation.

**Review by dann21 — APPROVED** [review]
LGTM

File: shared-lib/spiutils/src/compat/firmware.rs (hunk 1 above)
Comment (osenft) [reply shared-lib/spiutils/src/compat/firmware.rs]: Ha, ok. Fixed.

File: userspace/otpilot/src/spi_processor.rs (hunk 2 above)
Comment (osenft) [reply userspace/otpilot/src/spi_processor.rs]: Oh yeah, how did that happen? Thanks!

I'll need to run `rustfmt` on `spiutils`, `otpilot` and the new kernel code that I've written to make sure it conforms. I should have done that before and we should do that going forward. I'll send a follow up PR with just formatting changes.

**Pull request merged**